In [348]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [349]:
df_week = pd.read_csv("SAP_weekly_return_volatility.csv")
df_daily = pd.read_csv("SAP_weekly_return_volatility_detailed.csv")

In [350]:
df_week.head(5)

Year  Week_Number  mean_return  volatility  label
0  2019            0      -0.1050    3.594650      0
1  2019            1       0.6532    1.153409      1
2  2019            2       0.7326    1.198802      1
3  2019            3       0.0735    1.593999      0
4  2019            4      -0.3746    2.056743      0

In [351]:
df_daily.head(5)

High         Low        Open       Close     Volume  Adj Close  \
0   99.559998   98.339996   98.570000   99.180000   506300.0  94.896118   
1   96.830002   95.449997   96.730003   95.459999   807800.0  91.336800   
2   99.199997   96.910004   97.339996   98.739998  1038100.0  94.475128   
3  100.190002   98.760002   99.440002   99.709999   861600.0  95.403229   
4  101.480003  100.430000  101.370003  101.269997   548600.0  96.895844   

   Return        Date  Week_Number  Year  Day    Weekday  mean_return  \
0   0.000  2019-01-02            0  2019    2  Wednesday      -0.1050   
1  -3.751  2019-01-03            0  2019    3   Thursday      -0.1050   
2   3.436  2019-01-04            0  2019    4     Friday      -0.1050   
3   0.982  2019-01-07            1  2019    7     Monday       0.6532   
4   1.565  2019-01-08            1  2019    8    Tuesday       0.6532   

   volatility  
0    3.594650  
1    3.594650  
2    3.594650  
3    1.153409  
4    1.153409

#### 1. implement a "buy-and-hold" strategy for year 2 - you invest 100 on the first trading day (at the opening price) and sell at the last trading day at the (adjusted) closing price.

In [379]:
def buyhold(ret_list):
    ret = round(((ret_list / 100) + 1).cumprod() * 100,2)
    
    return ret

In [380]:
hold = buyhold(df_daily[(df_daily["Year"] == 2020)]["Return"])
hold.values[-1]

98.71

* Total Return using Hold and Buy strategy for 2020 returns \$98.71

#### 2. implement a trading strategy based on your labels for year 2 and compare the performance with the "buy-and-hold" strategy. Which strategy results in a larger amount at the end of the year?

In [376]:
def compute_daily_ret(d_frame1, amt):
    ret = round(((d_frame1["Return"] / 100) + 1).cumprod() * amt,2)

    return round(ret.values[-1],2)

In [377]:
def weekDaily(d_frame_week,d_frame_daily,lbl):
    total_ret = list()
    for i in range(len(d_frame_week)):
        if i > 0:
            year = d_frame_week.loc[i]["Year"]

            week = d_frame_week.loc[i]["Week_Number"]

            prv_week = d_frame_week.loc[i - 1]["Week_Number"]

            frame = d_frame_daily[(d_frame_daily["Week_Number"] == week)]

            if d_frame_week.loc[i - 1][lbl] == 0 and d_frame_week.loc[i][lbl] == 1:

                amount = total_ret[i - 1] 
                total_ret.append(compute_daily_ret(frame, amount))
                
            elif d_frame_week.loc[i - 1][lbl] == 1 and d_frame_week.loc[i][lbl] == 1:

                amount = total_ret[i - 1]
                total_ret.append(compute_daily_ret(frame, amount))
            else:
                total_ret.append(total_ret[i - 1])
        else:     
            total_ret.append(100)
    return total_ret

In [378]:
daily_ret = weekDaily(df_week[df_week["Year"] == 2020].reset_index(drop=True),\
                      df_daily[(df_daily["Year"] == 2020)].reset_index(drop=True),"label")
daily_ret[-1]

214.6

* My Strategy is better than the Buy and Hold strategy.
    * Buy and Hold = \$ 98.69
    * My Strategy = \$ 214.6